redis/src/cluster_async/mod.rs

```diff
@@ -355,18 +357,24 @@ where
     ) -> RedisResult<Self> {
         let connections =
             Self::create_initial_connections(initial_nodes, cluster_params.clone()).await?;
-        let mut connection = ClusterConnInner {
-            connections,
+        let connections_and_metadata = Arc::new(ConnectionsAndMetadata {
+            connections: Mutex::new(connections),
```

Comment:
I'm honestly not sure -- are we adding overhead by introducing the Mutex?

Reply:
I'm not 100% sure, but:
1. The mutex surrounds a Map, and most of the time it is only held for get/remove calls, which should be very fast, so even if the mutex is locked, the contention should be low.
2. The places where the mutex is held for long are functions that take &mut self, so they should be exclusive, and taking an unlocked mutex should have a negligible cost.

So as long as we're sure to either use it in exclusive context or for quick shared calls, this should be ok.

Reply:
also, it means that we don't leave the client in a weird state, like we do when we empty the connection map - ` let mut connections = mem::take(&mut self.connections);`